py/plugins/firebase/src/genkit/plugins/firebase/retriever.py

```diff
@@ -0,0 +1,138 @@
+    DistanceMeasure,
+    FirestoreRetrieverConfig,
+)
+
+
+class FirestoreRetriever:
+    """
+    Retrieves documents from Google Cloud Firestore using vector similarity search.
+
+    Attributes:
+        ai: An instance of the Genkit AI registry.
+        params: A FirestoreRetrieverConfig object containing the configuration
+            for the retriever
+        firestore_client: The initialized Firestore client from the configuration.
+    """
+
+    def __init__(self, ai: Genkit, params: FirestoreRetrieverConfig):
+        self.ai = ai
+        self.params = params
+        self.firestore_client = params.firestore_client
+        self._validate_config()
+
+    def _validate_config(self):
+        if not self.params.collection:
+            raise ValueError('Firestore Retriever config must include firestore collection name.')
+        if not self.params.vector_field:
+            raise ValueError('Firestore Retriever config must include vector field name.')
+        if not self.params.embedder:
+            raise ValueError('Firestore Retriever config must include embedder name.')
+        if not self.firestore_client:
+            raise ValueError('Firestore Retriever config must include firestore client.')
+
+    def build_query(self, request: RetrieverRequest):
+        """Builds  Retriever query from a request Query.
+        Fierstore Retriever only supports TextPart Request Content.
+
+        Args:
+            - request: a packed request for the retriever
+
+        Returns:
+            - query string for retriever
+        """
+        parts = request.query.content
+        if not parts or not isinstance(parts[0].root, TextPart):
+            return ''
+
+        return parts[0].root.text
+
+    def _to_content(self, doc_snapshot: DocumentSnapshot) -> list[Dict[str, str]]:
+        content_field = self.params.content_field
+        if callable(content_field):
+            return content_field(doc_snapshot)
+        else:
+            content = doc_snapshot.get(content_field)
+            return [{'text': content}] if content is not None else []
+
+    def _to_document(self, doc_snapshot: DocumentSnapshot) -> Document:
+        content = self._to_content(doc_snapshot)
+        metadata: Dict[str, Any] = {}
+        metadata_fields = self.params.metadata_fields
+        if metadata_fields:
+            for field in metadata_fields:
+                if field in doc_snapshot:
+                    metadata[field] = doc_snapshot.get(field)
+
+        return Document(content=content, metadata=metadata)
+
+    async def retrieve(self, request: RetrieverRequest, _: ActionRunContext) -> RetrieverResponse:
+        """
+        Retrieves documents from Firestore using native vector similarity search.
+
+        Args:
+            request: A RetrieverRequest Object
+        Returns:
+            A RetrieverResponse Object containing retrieved documents
+
+        """
+        query = self.build_query(request)
```

Comment:
Why build the query? `ai.embed` should accept the Document in RetrieverRequest...